### ISSUE TITLE
sccache seems to have no effect for Rust docker builds

### ISSUE THREAD

**Issue by oren0e** [mozilla/sccache#1160]
Hi,

According to what I understand from the readme here, in order to make sccache work with `cargo build --release` in Dockerfile I only need to install it using `cargo install sccache` and then set the environment variable `RUSTC_WRAPPER`, and then build the project normally. And when the build is triggered for the second time (when there is a cache present) - it should be much faster.

Well, this is not the case for me - maybe I'm doing something wrong. Here is my Dockerfile which I then trigger in Gitlab CI:
```docker
FROM rust:1.51 as builder

ENV NIGHTLY_VERSION nightly-2021-08-18
ENV LC_ALL C.UTF-8
ENV LANG C.UTF-8

# Use sccache
RUN cargo install sccache
ENV RUSTC_WRAPPER /usr/local/cargo/bin/sccache
ENV SCCACHE_CACHE_SIZE 90G
ENV SCCACHE_DIR /mnt/dispatcher-cache

RUN rustup override set $NIGHTLY_VERSION
RUN rustup component add rustfmt clippy --toolchain $NIGHTLY_VERSION-x86_64-unknown-linux-gnu
RUN cargo new --bin dispatcherr
WORKDIR ./dispatcherr
COPY ./Cargo.toml ./Cargo.toml
RUN cargo build --release
RUN rm src/*.rs

ADD . ./

RUN rm ./target/release/deps/dispatcherr*
RUN cargo build --release

FROM python:3.9-slim-buster

RUN apt-get update \
    && apt-get install -y ca-certificates tzdata \
    && rm -rf /var/lib/apt/lists/*

RUN pip install scottypy

EXPOSE 8000

RUN mkdir -p /home/dispatcherr
COPY --from=builder /dispatcherr/target/release/dispatcherr /home/dispatcherr
RUN chmod +x /home/dispatcherr

WORKDIR /home/dispatcherr

CMD ["./dispatcherr"]
```

When I run this stage, each time the compilation takes exactly the same (long) time. When I look at the directory where I defined to save the cache - it is empty. It seems like either I missed some steps from the readme or that sccache just does not work for some reason. 
I Would appreciate some help.

**Comment by mitchhentges** [comment]
Hmm, nothing is jumping out at me that might be incorrect.
It might be worth running `cargo -v build` to see what it's specifically calling - perhaps it's not invoking `sccache`.
You could even try manually running `/usr/local/cargo/bin/sccache rustc ...` (with a `rustc` command as printed by `cargo -v build`) to see if that generates files in the cache.

**Comment by jschwe** [comment]
I'm curious as to why you want to use sccache when _building_ the docker image. Usually you would want to use sccache when running inside the docker image. 

Also from what I'm seeing you are first building an empty project generated from `cargo new`, then copying some stuff over and then rebuilding. If the project files change, sccache can't have cached them.

**Comment by oren0e** [comment]
> I'm curious as to why you want to use sccache when _building_ the docker image. Usually you would want to use sccache when running inside the docker image.
> 
> Also from what I'm seeing you are first building an empty project generated from `cargo new`, then copying some stuff over and then rebuilding. If the project files change, sccache can't have cached them.

I want to save all the building information so in CI/CD later, when I change something that is not the code or dependencies - I won't have to build the image over and over again from scratch. In the Dockerfile above I first build only the dependencies and then rest of the project - if the code changes it is a different thing than when dependencies change.

**Comment by jschwe** [comment]
That is reasonable, but the dependencies are already cached by cargo (i.e. cargo won't rebuild them), and the result of that step is cached anyway by docker, so I don't see what could gained here by adding sccache to the mix.

**Comment by oren0e** [comment]
@jschwe you are correct, but at the end of the "docker build" step I purge everything from the CI runner not to leave any traces from the build as a standard procedure. I thought I would be able to use sccache to cache the results into a dedicated mounted place inside the server. You can argue that both options are identical.

**Comment by jschwe** [comment]
It appears as if you don't save and restore `/mnt/dispatcher-cache`, which would be needed to restore the local `sccache` cache. But maybe it would be easier to use the [`Redis`](https://github.com/mozilla/sccache#storage-options), `Memcached`, or one of the Cloud storage options to store the cache persistently?

**Comment by oren0e** [comment]
That is what I did not understand - I thought that merely defining the env var with `ENV SCCACHE_DIR /mnt/dispatcher-cache` is enough. Using Redis and the like does not improve performance much because of the downloading times.

**Comment by lcmgh** [comment]
@oren0e have you tried utilizing buildx cache functionality to persist `/mnt/dispatcher-cache` on the host?

**Comment by lcmgh** [comment]
This works for me

```Dockerfile
RUN --mount=type=cache,mode=0777,target=/home/root/.cache/sccache \
    cargo build --release && sccache --show-stats
```

What did not work was using a env. variable as target value.
```Dockerfile
RUN --mount=type=cache,mode=0777,target=$SCCACHE_DIR \
    cargo build --release && sccache --show-stats
```

**Comment by lcmgh** [comment]
Did you use buildx?
`docker buildx build .`

Have you set the wrapper variables in the Dockerfile?

```Dockerfile
ENV CARGO_HOME=/usr/local/cargo
ENV RUSTC_WRAPPER=/usr/local/cargo/bin/sccache
ENV SCCACHE_DIR=/home/root/.cache/sccache # must be same as the 'target' value for the mount cmd
```

**Comment by lcmgh** [comment]
hm. A workaround might be to use a remote cache such as S3.

**Comment by Xuanwo** [comment]
It is recommended to use cloud storage services such as S3 or GCS when using sccache in Docker or CI environments.
